src/lightning/data/datasets/iterable.py

```diff
@@ -0,0 +1,447 @@
+
+        # if state-dict was set, the curr chunk index was the global number across all shards
+        # --> divide it to get local number
+        if self._start_index_chunk and self._start_index_chunk == self._curr_chunk_index:
+            self._curr_chunk_index = math.ceil(self._curr_chunk_index // self._env.num_shards)
+
+    def _check_if_sharded(self) -> None:
+        """Raises a warning if the dataset is not sharded."""
+        if not self._local_chunks:
+            warnings.warn(
+                "Chunks have not been sharded yet. Call iter() on your dataset to ensure sharding is done correctly. "
+                "It won't recognize dataloader workers when manually calling it outside an actual dataloader."
+            )
+
+    def _check_dataset_end(self) -> None:
+        """Checks if the dataset has reached it's end or should be restarted."""
+        if self._curr_chunk_index >= len(self._local_chunks):
+            if self._wrap:
+                self._curr_chunk_index = 0
+            else:
+                raise StopIteration
+
+    def _shuffle_if_necessary(
+        self,
+        chunks: List[_Chunk],
+        first_chunk_index: int,
+        shuffle_chunk_order: bool = True,
+        shuffle_sample_order: bool = True,
+    ) -> List[_Chunk]:
+        """This shuffles the chunk-order and the order of samples within each chunk.
+
+        Args:
+            chunks: The chunks to optionally shuffle
+            first_chunk_index: The point to which the generator should be replayed
+            shuffle_chunk_order: Whether to shuffle the order of chunks
+            shuffle_sample_order: Whether to shuffle the order of samples within a chunk
+        """
+        # re-seed generator
+        if self._generator is not None and self._initial_generator_state is not None:
+            self._generator = self._generator.set_state(self._initial_generator_state)
+
+        # shuffle chunks if necessary
+        chunks = _Chunk(chunks, len(chunks))
+        if self._shuffle and shuffle_chunk_order:
+            chunks.shuffle(generator=self._generator)
+        # this is annoying but otherwise we cannot make sure the states are the same
+        elif self._shuffle:
+            _dummy_chunks = _Chunk(None, len(chunks._chunk_data))
+            _dummy_chunks.shuffle(generator=self._generator)
+        chunks = [chunks._chunk_data[i] for i in chunks]
+
+        if not shuffle_sample_order:
+            return chunks
+
+        # after shuffling all chunks -> fast forward to first_chunk_index
+        if self._shuffle and self._lazy_shuffle:
+            for _ in range(first_chunk_index):
+                chunk = _Chunk(None, chunk_size=self._chunk_size)
+                chunk.shuffle(generator=self._generator)
+        # shuffle samples within each chunk
+        elif self._shuffle:
+            chunks = [c.shuffle(generator=self._generator) for c in chunks]
+
+        return chunks
+
+
+class DataLoader(_DataLoader):
```

Comment:
The user might not know they have to use this dataloader. And it will probably work without notice if you use the regular dataloader from torch, right? Will we communicate this through docs?

Reply:
You can still use the dataloader for training, just resuming won't work unless you manually pass in the correct values for the seen samples and num_workers